### PR TITLE
fix collect and comprehension for NamedTuple and Dict

### LINKED ISSUE #725: bug in comprehension over dictionaries
It currently seems quite difficult to use `Dict`s with Zygote - at least I have trouble getting it to work at all. Here's a MWE where the `Dict` is obviously unnecessary but it illustrates my point nonetheless. I could be doing something wrong, however this is a common data structure used in a straightforward fashion.

```julia
using Zygote
w = randn(5)
function f(w)
    d = Dict{Int, Float64}(i => v for (i,v) in enumerate(w))
    sum([v for (_, v) in d])
end
gradient(f, w)
```
Error: **map is not defined on dictionaries**. 
Using iterator syntax for the sum instead of the list comprehension syntax gives: **MethodError: no method matching getindex(::Dict{Any,Any})**, which probably makes the issue similar to #408.

### PR DESCRIPTION
**Pull request by CarloLucibello**
supersedes #1064

fix #725

### CODE REVIEW

File: test/lib/array.jl [hunk 1]
```diff
@@ -18,3 +18,46 @@ test_rrule(ZygoteRuleConfig(), x->sum(sin, Diagonal(x)), rand(3); rrule_f=rrule_
...
+        g = gradient(d -> sum(v^2 for (_,v) in collect(d)), d)[1]
+        @test g isa Dict{Int,Int}
+        @test g == Dict(1 => 10, 2 => 12)
+    end
+
+    @testset "NamedTuple" begin
+        t = (a=1, b=2)
+        g = gradient(d -> sum(x^2 for x in collect(d)), t)[1]
+        @test g isa NamedTuple{(:a, :b), Tuple{Float64, Float64}}
+        @test  g == (a = 2, b = 4)
```
Comment (ToucheSir) [inline test/lib/array.jl:40]:
```suggestion
        @test g === (a = 2.0, b = 4.0)
```
I believe this is sufficient to capture both tests

**Review by ToucheSir — COMMENTED** [review]
Do we already have a test for `collect(::Tuple)`? Otherwise LGTM!

**Comment by CarloLucibello** [comment]
> Do we already have a test for collect(::Tuple)? Otherwise LGTM!

I don't know, tests are so scattered around, but I added one just in case

**Review by mcabbott — COMMENTED** [review]
Some things I noticed, which this PR doesn't in fact change, but may deserve follow-up:

File: src/lib/array.jl [hunk 2]
```diff
@@ -8,8 +8,26 @@ using Distributed: pmap, AbstractWorkerPool
...
+  collect(x), collect_tuple_pullback
+end
+
+@adjoint function collect(x::NamedTuple{names}) where names
+  collect_namedtuple_pullback(dy) = (NamedTuple{names}(Tuple(dy)),) 
+  collect(x), collect_namedtuple_pullback
+end
+
+@adjoint function collect(x::AbstractArray)
+  collect_array_pullback(dy) = (dy,)
```
[the remark this answers is not in the document]
Comment (CarloLucibello) [reply src/lib/array.jl:22]: I'm not sure I understand the question. Is that about 
```julia
julia> x = cu(rand(5));
5-element CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}:
 0.29323888
 0.5233124
 0.34484053
 0.5885852
 0.77851206

julia> gradient(x -> sum(rand(5) .* collect(x)), x)[1]
5-element Vector{Float32}:
 0.048657022
 0.7181719
 0.4261651
 0.80847126
 0.31513062
```
Should we return a CuArray gradient instead?

File: src/lib/array.jl [hunk 3]
```diff
@@ -8,8 +8,26 @@ using Distributed: pmap, AbstractWorkerPool
 
 @adjoint copy(x::AbstractArray) = copy(x), ȳ -> (ȳ,)
 
-@adjoint collect(x::Tuple) = collect(x), dy -> (Tuple(dy),)
-@adjoint collect(x::AbstractArray) = collect(x), dy -> (dy,)
+@adjoint function collect(x::Tuple)
+  collect_tuple_pullback(dy) = (Tuple(dy),)
```
[the remark this answers is not in the document]
Comment (CarloLucibello) [reply src/lib/array.jl:12]: so something like
```julia
n = length(x)
collect_tuple_pullback(dy) = ntuple(i -> dy[i], n)  
``
`
?

[the remark this answers is not in the document]
Comment (mcabbott) [reply src/lib/array.jl:12]: Yes. Or perhaps
```julia
n = Val(length(x))
collect_tuple_pullback(dy0) = (dy=unthunk(dy0); ntuple(i -> dy[i], n)) 
```
to ensure it un-thunks at most once (although maybe the CR bridge handles that, I forget) and to make sure the length is in the type domain (although maybe the compiler figures it out here, not sure.)

File: src/lib/array.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (mcabbott) [reply src/lib/array.jl:22]: Yes, ideally we should. 

But perhaps not the point of this PR, sorry.

[the remark this answers is not in the document]
Comment (ToucheSir) [reply src/lib/array.jl:22]: In the spirit of trying to do something reasonable but not trying too hard...
```julia
@adjoint function collect(x::AbstractArray)
  collect_array_pullback(dy) = (dy,) # nothing, etc.
  collect_array_pullback(dy::AbstractArray) = (_project(x, dy),)
  collect(x), collect_array_pullback
end
```

File: src/lib/array.jl [hunk 4]
```diff
@@ -8,8 +8,26 @@ using Distributed: pmap, AbstractWorkerPool
...
+  collect_namedtuple_pullback(dy) = (NamedTuple{names}(Tuple(dy)),) 
+  collect(x), collect_namedtuple_pullback
+end
+
+@adjoint function collect(x::AbstractArray)
+  collect_array_pullback(dy) = (dy,)
+  collect(x), collect_array_pullback
+end
+
+@adjoint function collect(d::AbstractDict)
```
Comment (ToucheSir) [inline src/lib/array.jl:26]: I just noticed `reconstruct_if_dict` always returns a `Dict`. In that light, tightening this signature to 
```suggestion
@adjoint function collect(d::Dict)
```
feels like a good idea. No need to think about what would happen with IdDicts or more exotic types.

File: src/lib/array.jl (hunk 3 above)
[the remark this answers is not in the document]
Comment (ToucheSir) [reply src/lib/array.jl:12]: The bridge does indeed unthunk eagerly before calling the `@adjoint` pullback.

File: src/lib/array.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (CarloLucibello) [reply src/lib/array.jl:22]: stuff for another PR

**Comment by CarloLucibello** [comment]
merge?

**Review by ToucheSir — APPROVED** [review]

**Pull request merged**